File: abstract-factory-design-pattern/C++/ui.cpp
#include <string>
#include <iostream>

using namespace std;

class IButton {
    public:
        virtual void press() = 0;
};

class ITextBox {
    public:
        virtual void setText() = 0;
};

class WindowsButton : public IButton {
    public:
        void press() {
            cout << "Windows button pressed." << endl;
        }
};

class MacButton : public IButton {
    public:
        void press() {
            cout << "Mac button pressed." << endl;
        }
};

class WindowsTextBox : public ITextBox {
    public:
        void setText() {
            cout << "Setting text in Windows TextBox." << endl;
        }
};

class MacTextBox : public ITextBox {
    public:
        void setText() {
            cout << "Setting text in Mac TextBox." << endl;
        }
};

class IFactory {
    public:
        virtual IButton* createButton() = 0;
        virtual ITextBox* createTextBox() = 0;
};

class WindowsFactory : public IFactory {
    public:
        IButton* createButton() {
            return new WindowsButton();
        }
        ITextBox* createTextBox() {
            return new WindowsTextBox();
        }
};

class MacFactory : public IFactory {
    public:
        IButton* createButton() {
            return new MacButton();
        }
        ITextBox* createTextBox() {
            return new MacTextBox();
        }
};

class GUIAbstractFactory {
    public:
        static IFactory* createFactory(string osType){
            if(osType == "mac"){
                return new MacFactory();
            }
            else if(osType == "windows") {
                return new WindowsFactory();
            }
            else{
                return nullptr;
            }
        }
};

int main() {

    cout << "Enter Machine OS: \n";
    string osType;
    cin >> osType;

    IFactory* factory = GUIAbstractFactory::createFactory(osType);

    IButton* button = factory->createButton();
    ITextBox* textBox = factory->createTextBox();

    button->press();
    textBox->setText();

    return 0;
}

File: abstract-factory-design-pattern/Python/ui.py
class IButton:
    def press(self):
        pass

class ITextBox:
    def set_text(self):
        pass
    
class WindowsButton(IButton):
    def press(self):
        print("Windows button pressed.")

class MacButton(IButton):
    def press(self):
        print("Mac button pressed.")

class WindowsTextBox(ITextBox):
    def set_text(self):
        print("Setting text in Windows TextBox.")

class MacTextBox(ITextBox):
    def set_text(self):
        print("Setting text in Mac TextBox.")

class IFactory:
    def createButton(self):
        pass
    def createTextBox(self):
        pass
    
class WindowsFactory(IFactory):
    def createButton(self):
        return WindowsButton()
    def createTextBox(self):
        return WindowsTextBox()

class MacFactory(IFactory):
    def createButton(self):
        return MacButton()
    def createTextBox(self):
        return MacTextBox()

class GUIAbstractFactory():
    @staticmethod
    def createFactory(os_type):
        if os_type == "windows":
            return WindowsFactory()
        elif os_type == "mac":
            return MacFactory()
        else:
            return None
        
def main():
    os_type = input("Enter Machine OS: ")
    
    factory = GUIAbstractFactory.createFactory(os_type)
    
    if factory is not None:
        button = factory.createButton()
        button.press()
        text_box = factory.createTextBox()
        text_box.set_text()
    else:
        print("Invalid OS Type!!")

if __name__ == "__main__":
    main()